File: internal/search/search_test.go
package search_test

import (
	"testing"

	"github.com/adriankarlen/storyquery/internal/manifest"
	"github.com/adriankarlen/storyquery/internal/search"
)

func sampleComponents() map[string]manifest.Component {
	return map[string]manifest.Component{
		"components-buttons-mainbutton": {
			ID: "components-buttons-mainbutton", Name: "MainButton",
		},
		"components-buttons-navigationbutton": {
			ID: "components-buttons-navigationbutton", Name: "NavigationButton",
		},
		"components-alert": {ID: "components-alert", Name: "Alert"},
	}
}

func TestComponentsRanking(t *testing.T) {
	comps := sampleComponents()
	tests := []struct {
		name   string
		term   string
		wantID string // expected top match id, "" = no match
	}{
		{"exact name", "MainButton", "components-buttons-mainbutton"},
		{"exact name lowercase", "mainbutton", "components-buttons-mainbutton"},
		{"exact id", "components-alert", "components-alert"},
		{"name prefix", "Main", "components-buttons-mainbutton"},
		{"id substring", "buttons", "components-buttons-mainbutton"},
		{"fuzzy", "navbtn", "components-buttons-navigationbutton"},
		{"no match", "zzzzz", ""},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := search.Components(comps, tt.term, 0)
			if tt.wantID == "" {
				if len(got) != 0 {
					t.Fatalf("expected no matches, got %d", len(got))
				}
				return
			}
			if len(got) == 0 {
				t.Fatalf("expected a match for %q", tt.term)
			}
			if got[0].Component.ID != tt.wantID {
				t.Errorf("top match = %q, want %q", got[0].Component.ID, tt.wantID)
			}
		})
	}
}

func TestComponentsLimit(t *testing.T) {
	comps := sampleComponents()
	got := search.Components(comps, "button", 1)
	if len(got) != 1 {
		t.Errorf("limit 1: got %d", len(got))
	}
}

func TestBestComponent(t *testing.T) {
	comps := sampleComponents()

	m, ok, ambiguous := search.BestComponent(comps, "MainButton")
	if !ok || ambiguous {
		t.Fatalf("exact name: ok=%v ambiguous=%v", ok, ambiguous)
	}
	if m.Component.ID != "components-buttons-mainbutton" {
		t.Errorf("got %q", m.Component.ID)
	}

	if _, ok, _ := search.BestComponent(comps, "zzzz"); ok {
		t.Error("expected no match")
	}
}

func TestDocsRanking(t *testing.T) {
	docs := map[string]manifest.Doc{
		"developers-getting-started--docs":        {ID: "developers-getting-started--docs", Title: "Developers/Getting started", Name: "Docs"},
		"components-alert-guidelines-usage--docs": {ID: "components-alert-guidelines-usage--docs", Title: "Components/Alert/Guidelines/Usage", Content: "use alerts sparingly"},
	}
	got := search.Docs(docs, "alert", 0)
	if len(got) == 0 || got[0].Doc.ID != "components-alert-guidelines-usage--docs" {
		t.Fatalf("expected alert guideline top match, got %+v", got)
	}

	// Content-only match.
	got = search.Docs(docs, "sparingly", 0)
	if len(got) != 1 {
		t.Errorf("expected 1 content match, got %d", len(got))
	}
}

func FuzzComponentsQuery(f *testing.F) {
	comps := sampleComponents()
	for _, seed := range []string{"MainButton", "", "btn", "components-alert", "界"} {
		f.Add(seed)
	}
	f.Fuzz(func(t *testing.T, term string) {
		// Must never panic and must return a sorted, non-increasing score list.
		got := search.Components(comps, term, 0)
		for i := 1; i < len(got); i++ {
			if got[i-1].Score < got[i].Score {
				t.Fatalf("scores not sorted descending: %d < %d", got[i-1].Score, got[i].Score)
			}
		}
	})
}

File: test/search.test.ts
import { describe, expect, it } from "vitest";

import type { Component, Doc } from "../src/manifest/types.js";
import { bestComponent, fuzzyScore, searchComponents, searchDocs } from "../src/search.js";

function comp(id: string, name: string): Component {
  return { id, name };
}

const comps: Record<string, Component> = {
  a: comp("components-button", "Button"),
  b: comp("components-iconbutton", "IconButton"),
  c: comp("components-alert", "Alert"),
};

describe("searchComponents", () => {
  it("ranks exact name highest", () => {
    const res = searchComponents(comps, "Button", 0);
    expect(res[0]?.component.name).toBe("Button");
    expect(res[0]?.score).toBe(1000);
  });

  it("exact id beats prefix", () => {
    const res = searchComponents(comps, "components-alert", 0);
    expect(res[0]?.component.id).toBe("components-alert");
    expect(res[0]?.score).toBe(900);
  });

  it("name prefix outranks substring", () => {
    const res = searchComponents(comps, "but", 0);
    // "Button" starts with "but" -> prefix; IconButton only contains it.
    expect(res[0]?.component.name).toBe("Button");
  });

  it("respects limit", () => {
    expect(searchComponents(comps, "components", 1)).toHaveLength(1);
    expect(searchComponents(comps, "components", 0).length).toBeGreaterThan(1);
  });

  it("empty term matches nothing", () => {
    expect(searchComponents(comps, "", 0)).toHaveLength(0);
  });
});

describe("bestComponent", () => {
  it("returns found for a unique exact match", () => {
    const r = bestComponent(comps, "Button");
    expect(r.kind).toBe("found");
  });

  it("returns ambiguous when top scores tie", () => {
    const r = bestComponent(comps, "components");
    expect(r.kind).toBe("ambiguous");
  });

  it("returns none for no match", () => {
    expect(bestComponent(comps, "zzzz").kind).toBe("none");
  });
});

describe("searchDocs", () => {
  const docs: Record<string, Doc> = {
    a: { id: "intro--docs", title: "Introduction", content: "welcome here" },
    b: { id: "tokens--docs", title: "Design Tokens", content: "colors" },
  };
  it("matches on title", () => {
    const res = searchDocs(docs, "introduction", 0);
    expect(res[0]?.doc.title).toBe("Introduction");
  });
  it("matches weakly on content", () => {
    const res = searchDocs(docs, "colors", 0);
    expect(res[0]?.doc.id).toBe("tokens--docs");
    expect(res[0]?.score).toBe(50);
  });
});

describe("fuzzyScore", () => {
  it("matches a subsequence", () => {
    expect(fuzzyScore("button", "btn")).not.toBeNull();
  });
  it("rejects a non-subsequence", () => {
    expect(fuzzyScore("button", "xyz")).toBeNull();
  });
  it("rewards density", () => {
    const dense = fuzzyScore("button", "but")!;
    const sparse = fuzzyScore("b-u-t-x", "but")!;
    expect(dense).toBeGreaterThan(sparse);
  });

  // Fuzz-style: random inputs must never throw and scores stay non-negative.
  it("never throws and scores >= 0 for random input", () => {
    const chars = "abcdefg-0 ";
    const rand = (n: number) =>
      Array.from({ length: n }, () => chars[Math.floor(Math.random() * chars.length)]).join("");
    for (let i = 0; i < 2000; i++) {
      const s = rand(Math.floor(Math.random() * 12));
      const q = rand(Math.floor(Math.random() * 5));
      const r = fuzzyScore(s, q);
      if (r !== null) expect(r).toBeGreaterThanOrEqual(0);
    }
  });
});
